interfaces/apparmor/apparmor.go

```diff
@@ -0,0 +1,94 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package apparmor contains primitives for working with apparmor.
+//
+// References:
+//  - http://wiki.apparmor.net/index.php/Kernel_interfaces
+//  - http://apparmor.wiki.kernel.org/
+//  - http://manpages.ubuntu.com/manpages/xenial/en/man7/apparmor.7.html
+package apparmor
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// LoadProfile loads an apparmor profile from the given file.
+//
+// If no such profile was previously loaded then it is simply added to the kernel.
+// If there was a profile with the same name before, that profile is replaced.
+func LoadProfile(fname string) error {
+	return exec.Command("apparmor_parser", "--replace", fname).Run()
+}
+
+// Profile contains the name and mode of an apparmor profile loaded into the kernel.
+type Profile struct {
+	// Name of the profile. This is is either full path of the executable or an
+	// arbitrary string without spaces.
+	Name string
+	// Mode is either "enforcing" or "complaining".
+	Mode string
+}
+
+// Unload removes a profile from the running kernel.
+//
+// The operation is done with: apparmor_parser --remove $name
+func (profile *Profile) Unload() error {
+	return exec.Command("apparmor_parser", "--remove", profile.Name).Run()
+}
+
+// profilesPath contains information about the currently loaded apparmor profiles.
+const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
+
+var profilesPath = realProfilesPath
+
+// LoadedProfiles interrogates the kernel and returns a list of loaded apparmor profiles.
+//
+// Snappy manages apparmor profiles named *.snap. Other profiles might exist on
```

Comment:
1. The launcher is expecting a profile name. Adding '.snap' to the
profile name means we can either change the launcher to append it or we
can change the /snaps/bin generation and the systemd unit generation. I
prefer the latter.

ZK: I'll patch the launcher generator to use .snap suffix.

2. profileAttachForApp() needs to change to append .snap

ZK: yep, I'll write one, authoritative function that computes a profile name for snap + app and use it everywhere.

3. varsForApp() needs to change so that encoded '.snap' (_2esnap)
appears in APP_ID_DBUS

ZK: Noted, I'll make that happen in the other branch ( #617 )

I prefer you abstract this out into a 'profileNameForApp()' (or
similar) that the above 3 points can use rather than sprinkling .snap
in various places.

ZK: Yeah, I already planned to do this while fixing hw-assign. I'll make that happen separately so that we can hunt-kill all remaining duplications of the code that computes this.

Moving forward, we might have to consider '.snap' as part of the label
for new policy when writing rules for new interfaces (eg, that use
dbus, unix, signal and ptrace and in the future, network).

ZK: I'm not sure I understand this completely. Yes, .snap suffix will show up in udev rules (because those are raw snippets) but apparmor and seccomp don't really care about the name (as far as the content of the actual profiles is concerned). Please correct me, I'm sure I'm missing something.

That should be it but I maintain that adding .snap adds complexity
where I think it shouldn't necessarily be and the implementation is
prone to bugs.

ZK: The motivation to add .snap is to ensure that we can implement Ensure() reliably. Ensure is supposed to inspect the state of the system, compute a desired state of the system and fix any issues. One aspects of doing that is to ensure we get rid of unused in-kernel apparmor profiles. Since the snap manager can remove a snap without telling us we need to notice that there are files and profiles in the air that don't have a corresponding snap and act on them accordingly. I agree that there's a potential clash with 3rd party environments that use .snap suffix somewhere but I think this is not unreasonable as an assumption to take, that we kind-of manage that pseudo naming space (not a kernel namespace). If you know of another method that we could use to do this then I'd love to know what that is.